### PR TITLE
[WIP] Streamlined Scope

### PR DESCRIPTION
**Pull request by quartzjer**
There are two significant things to note in this PR:
1. Renaming of the `derive` step and "derivation" to the `prove` step and generating proofs, to be more consistent with the usage of "proofs" elsewhere in the draft.
2. Removing all language regarding the mapping of what values are in which payload slot, leaving this to the application or a companion spec.

The result is a simpler spec that focuses strictly on the cryptographic container with pluggable proof algorithms.

### CODE REVIEW

File: draft-jmiller-json-web-proof.md [hunk 1]
```diff
@@ -37,7 +37,7 @@ organization = "Ping Identity"
...
 
-This document adds to the JOSE family by standardizing a new container format very similar in purpose and design to a JWS, called a _JSON Web Proof (JWP)_.  It adds support for the new algorithms by way of containing multiple individual payloads instead of a singular one as well as an additional derivation step to apply the privacy preserving selection and computation.
+This document adds to the JOSE family by standardizing a new container format very similar in purpose and design to a JWS, called a _JSON Web Proof (JWP)_.  It adds support for the new algorithms by way of containing multiple individual payloads instead of a singular one as well as an additional proof-generation step to apply the privacy preserving selection and computation.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:40]: I am not sure this change is good. 

`proof-generation` is relevant to JWS/JWT, but `proof-derivation` is a new concept for JOSE  imo.

File: draft-jmiller-json-web-proof.md [hunk 2]
```diff
@@ -47,16 +47,17 @@ The JOSE specifications are very widely deployed and well supported technology c
 
 With these new use cases, there is an increased focus on adopting privacy-protecting cryptographic primitives.  While such primitives are still an active area of academic and applied research, the leading candidates introduce new patterns that are not currently supported by JOSE.  These new patterns are largely focused on two areas: supporting selective disclosure when presenting a credential and minimizing correlation through the use of Zero-Knowledge Proofs (ZKPs), instead of traditional signatures.
 
-There are a growing number of these cryptographic primitives that support selective disclosure while protecting privacy across multiple presentations.  Examples that have already been or are being deployed in the context of Verifiable Credentials are:
+There are a growing number of these cryptographic primitives which support selective disclosure while protecting privacy across multiple presentations.  Examples that have already been or are being deployed in the context of Verifiable Credentials are:
 
 * [CL Signatures](https://eprint.iacr.org/2012/562.pdf)
 * [IDEMIX](http://www.zurich.ibm.com/idemix)
 * [BBS+](https://github.com/mattrglobal/bbs-signatures)
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:54]:
```suggestion
* [BBS+](https://github.com/mattrglobal/bbs-signatures)
* [MerkleDisclosureProof2021](https://github.com/transmute-industries/merkle-disclosure-proof-2021)
```

File: draft-jmiller-json-web-proof.md [hunk 3]
```diff
@@ -47,16 +47,17 @@ The JOSE specifications are very widely deployed and well supported technology c
...
 * [Mercural Signatures](https://eprint.iacr.org/2020/979)
 * [PS Signatures](https://eprint.iacr.org/2015/525.pdf)
 * [U-Prove](https://www.microsoft.com/en-us/research/project/u-prove/)
+* [Spartan](https://github.com/microsoft/Spartan)
 
-All of these follow the same pattern of taking multiple claims (a.k.a., "attributes" or "messages" in the literature) and binding them together into an issued credential.  These are then later securely one-way transformed into a presentation, revealing potentially only a subset of the original claims as required.
+All of these follow the same pattern of taking multiple claims (a.k.a., "attributes" or "messages" in the literature) and binding them together into an issued credential.  These are then later securely one-way transformed into a presentation, revealing potentially only a subset of the original claims as required or just proofs of the values.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:60]: I know what you mean here, but potentially we might use language that overlaps with https://csrc.nist.gov/Projects/pec/zkproof

Something to the effect of "...only a subset of the original claims" or  "Proofs of knowledge"  for the original claims.

File: draft-jmiller-json-web-proof.md [hunk 4]
```diff
@@ -70,49 +71,25 @@ when, and only when, they appear in all capitals, as shown here.
...
 
-In addition to the JWS `sign` and `verify` interactions, JWP also importantly adds a `derive` processing step for interaction with the algorithm to perform the selective disclosure and privacy preserving transformations.  This allows for multi-party interactions where a credential is issued from one party, derived by an intermediary party, then presented to another verifying party.  While `sign` only occurs once on a JWP, `derive` and `verify` may be repeated if supported by the algorithm.
+In addition to the JWS `sign` and `verify` interactions, JWP also importantly adds a `prove` processing step for interaction with the algorithm to perform the selective disclosure and privacy preserving transformations.  This allows for multi-party interactions where a credential is issued from one party, derived by the prooving party, then presented to another verifying party.  While `sign` only occurs once to create a JWP, the `prove` and `verify` steps may be safely repeated when supported by the algorithm.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:74]: I would prefer to see JWP support interfaces like:

`create/generate` - construct and secure document and proof.
`derive/prove` - transform a document and proof in such a way that they remain verifiable while's enhancing privacy.
`verify` - verify a document and proof.

I'm not sure its wise to limit `sign/create` to happening only once.... consider multi-signature scenarios where material is collected from parties over time.

We could potential address those scenarios by thinking of `sign` as `contribute to proof` where it might happen once, or a few times.

File: draft-jmiller-json-web-proof.md [hunk 5]
```diff
@@ -70,49 +71,25 @@ when, and only when, they appear in all capitals, as shown here.
...
 
-# JWP Format
-
-A JWP contains multiple specific payloads, which are always represented within an ordered array. These payloads have the same processing rules applied as they would within JWS.
-
-The individual payloads are often meant to be composed into a single credential, and as such, are most commonly not wholly indpendent from one another, but rather serve as facets of a whole. In keeping with the nomenclature of JWTs, such payloads are said to each represent one or more _claims_.
-
-An individual payload may contain structured information, such as a JSON document representing multiple claims. Another payload might represent a single JWT claim, such as a binary profile image. Other payloads may represent cryptographic values for supporting various proofs, which might also be interpreted as claims.
+Algorithm definitions that support JWPs will be done in separate companion specifications - just as the [JSON Web Algorithms] [@RFC7518] specification does for JWS and JWE [@RFC7516].
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:78]: <3 this feels like an excellent way of protecting JWP from over coupling to a particular approach, but we will need to remain vigilant.

File: draft-jmiller-json-web-proof.md [hunk 6]
```diff
@@ -70,49 +71,25 @@ when, and only when, they appear in all capitals, as shown here.
...
+# JWP Format
 
-The key used for a JWP that is identified either by its header or by context MUST have an associated ordered array of payload definitions.  If defined by a JWK, this specification registers a `payloads` parameter with an array value, containing one object for each payload with its associated header values.
+A JWP always contains a protected header along with one or more specific payloads.  The payloads are always serialized and processed as an ordered array.
 
-If the JWK includes a `use` parameter this specification registers a `prf` (proof) value that MUST be used in order for the JWK to be valid for a JWP.
+An individual payload may contain structured information such as a JSON document or be a simple value such as a number, string, or even a binary image.  Some algorithms may support payload values that are cryptographic values such as elliptic curve points or blinded secrets.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:84]: This feels like too many legal values for a payload....

In typescript this type would look like this:

```
type Payload: string | number | Buffer | any; // yikes any? really?
```

This will cause security issues later.

IMO, payload should be JSON... if you can't serialize or deserialize JSON (IoT / CBOR/ pure binary use cases)....you probably won't like base64url encoding either... 

Alternatively, we might consider CBOR instead of JSON, but I really don't like the payload not having a single know type.

File: draft-jmiller-json-web-proof.md [hunk 7]
```diff
@@ -70,49 +71,25 @@ when, and only when, they appear in all capitals, as shown here.
...
-      {"claims":["birthdate"]},
-      {"claims":["age"], "cty":"hashchain-commitment"},
-      {"claims":["profile_pic"], "cty":"image/png"}
-   ]
-  }
-```
+Although there are multiple payloads, the protected header still represents the JWP as a whole.
 
-## Protected Header
+It is recommended that payload-specific information is not included in the header and is handled outside of the cryptographic envelope.  This is to minimize any correlatable signals in the metadata, to prevent a verifier from categorizing based on header changes that may varry between multiple JWPs.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:90]: I agree with this, but we should beware of scenarios where time information MUST be disclosed...

proving I have a DL, without proving it's not expired could be a problem... pushing that information to the payload could cause interop issues, similar to `nbf`, `iat`....

File: draft-jmiller-json-web-proof.md [hunk 8]
```diff
@@ -121,44 +98,27 @@ For example:
 }
 ```
 
+Every JWP algorithm must include a digest method that is used to generate a hash of the base64url serialized protected header.  The protected header cannot be selectively disclosed and the digest value MUST be included in all proof values.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:101]: This sounds to me like https://datatracker.ietf.org/doc/html/rfc7638 but applying to JWP protected headers.

We should beware of serialization ordering issues with objects.... 

This digest would function as a correlation identifier for any derived proof.

File: draft-jmiller-json-web-proof.md [hunk 9]
```diff
@@ -121,44 +98,27 @@ For example:
...
+
 ## Payloads
 
-Payloads are always represented as an ordered array.  Each payload's content type is either known by context or specified in that payload's associated header.  The default content type is `application/json`, with the JSON object values being JWT claims.
+Payloads are always represented as an ordered array.  The mapping of which value is in which payload slot is out of scope of this specification.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:105]: We don't have a definition of `value` nor a single type of payload, which makes this sentence very hard to understand....

In my preferred world where the payloads are JSON, this might read as:

The order and construction of specific payload objects prior to JSON serialization is out of scope for this specification as is the order in which individual payloads are assembled into the payloads.... the sentence above is trying to do too much at once...

we need to address both:

```
type Payload: JSON | ANY

type Payloads: OrderedArray<Payload>
```

both are out of scope.

File: draft-jmiller-json-web-proof.md [hunk 10]
```diff
@@ -121,44 +98,27 @@ For example:
...
-{
-    "email":"janedoe@example.com"
-}
-```
-```json
-{
-    "birthdate":"0000-03-22"
-}
-```
+In order to support ZKPs, individual payloads cannot be serialized before they are passed into an algorithm implementation.  This enables the algorithms to accept and internally encode elliptic curve points, blinded values, plain numbers, membership keys, etc.  Implementations are therefore required to provide optional arguments for each payload such that the application can utilize these capabilities as needed.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:107]: I am not sure I agree with this... how you internally represent EC points is not relevant to a JSON Data Format spec... 

Might be better to say: encoding requirements for libraries that produce JWPs are out of scope, as are defining JSON serializations for ZKP parameters or options.... however I suspect that will have to happen somewhere...

File: draft-jmiller-json-web-proof.md [hunk 11]
```diff
@@ -121,44 +98,27 @@ For example:
...
 
+When selective disclosure preferences are applied, any one or more payloads may be hidden.  The position of other payloads does not change due to any preceeding ones being hidden - the resulting array will simply be sparse without the hidden payloads.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:109]: This sentence does not make sense to me... are you meaning to say that the length of `payload: Array<ANY>` is constant / final and hidding values is like doing this:

items[disclosureIndex'] = 'hidden' ?

File: draft-jmiller-json-web-proof.md [hunk 12]
```diff
@@ -121,44 +98,27 @@ For example:
...
 
-All proofs MUST include integrity protection of the JWP's base64url encoded header value.  This value cannot be hidden and the protection MUST be included in the proof even after derivation.
-
-# Derivation
-
-When selective disclosure preferences are applied to a derivation, any one or more payloads may be hidden.  The position of other payloads does not change due to any proceeding ones being hidden; the resulting array will simply be sparse, with the hidden payloads being replaced by the value `null`.
+Implementations will also be required to provide optional arguments for each payload as input into the `prove` step.  These arguments can be used for generating predicate proofs, linking options, etc.
 
-Algorithms SHOULD generate a new un-correlatable proof value when a JWP is derived.  A JWP may also be single-use, where correlation across multiple derivations is not a factor.
+Algorithms SHOULD generate a new un-correlatable proof value during the `prove` step.  A JWP may also be single-use, where correlation across multiple derivations is not a factor.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:117]: this is not possible if disclosure of the original header digest is required.

File: draft-jmiller-json-web-proof.md [hunk 13]
```diff
@@ -121,44 +98,27 @@ For example:
...
 
 # Serializations
 
-Each payload MUST be base64url encoded when preparing it to be serialized.  The header and proof are also individually base64url encoded.
+Each disclosed payload MUST be base64url encoded when preparing it to be serialized.  The header and proof are also individually base64url encoded.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:121]: As much as I love base64url... it kinda sucks that `-` breaks highlighting... I take it this is one of those cases where the old way of doing things is probably worth preserving more than the improvement, buts its worth noting that base64url inflates data, has no built in error correction, is not "double clickable" and allows multiple serializations to deserialize to the same buffer.

These issues might not be worth deviating from the convention of just using bas64url, but they are worth noting.

**Review by OR13 — COMMENTED** [review]
This PR is awesome, I left some comments, none of my comments are blocking as this is DIF not W3C or IETF.

**Comment by quartzjer** [comment]
Thank you @OR13, this is some excellent feedback! 

I've got some work to do incorporating it, which might take me a week or so since I'll be out most of this coming week.

File: draft-jmiller-json-web-proof.md (hunk 6 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:84]: JWS supports an unlimited amount of legal values for a payload - in that it's simply signing an array of bytes.  Similarly, the JWP layer should know nothing about the content of ifs payloads, other than their representations as arrays of bytes.

File: draft-jmiller-json-web-proof.md (hunk 7 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:90]: We should talk about "MUST be disclosed" payloads on the next DIF Crypto JWP call.  It's a great topic.  That said, I don't think this PR should block on solving that.

File: draft-jmiller-json-web-proof.md (hunk 9 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:105]: A spec layered about JWP where payloads are JSON should absolutely define this.  We'd tentatively started calling this JSON Selective Disclosure Token (JSDT), I believe.  The beauty of this PR is that it separates concerns, in the same way that concerns are separated between JWS and JWT.

File: draft-jmiller-json-web-proof.md [hunk 14]
```diff
@@ -47,16 +47,17 @@ The JOSE specifications are very widely deployed and well supported technology c
 
 With these new use cases, there is an increased focus on adopting privacy-protecting cryptographic primitives.  While such primitives are still an active area of academic and applied research, the leading candidates introduce new patterns that are not currently supported by JOSE.  These new patterns are largely focused on two areas: supporting selective disclosure when presenting a credential and minimizing correlation through the use of Zero-Knowledge Proofs (ZKPs), instead of traditional signatures.
 
-There are a growing number of these cryptographic primitives that support selective disclosure while protecting privacy across multiple presentations.  Examples that have already been or are being deployed in the context of Verifiable Credentials are:
+There are a growing number of these cryptographic primitives which support selective disclosure while protecting privacy across multiple presentations.  Examples that have already been or are being deployed in the context of Verifiable Credentials are:
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:50]: "that" is grammatically correct here.  Please revert the change to "which".

File: draft-jmiller-json-web-proof.md (hunk 8 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:101]: Another great topic for the next call...

File: draft-jmiller-json-web-proof.md (hunk 1 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:40]: It seems to me that there's both proof generation and proof derivation steps, as well as a proof verification step.  Which were you intending to talk about here (or maybe both of the first two)?

File: draft-jmiller-json-web-proof.md (hunk 10 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:107]: This has to be in scope somewhere.  It may be the algorithms spec.  It may also make its way into the JSDT spec.  Clean engineering of the layers will be key.

File: draft-jmiller-json-web-proof.md (hunk 12 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:117]: Does Derivation move to the JSDT spec?  We may still want to at least informatively say that applications of JWP will do derivation.

File: draft-jmiller-json-web-proof.md (hunk 13 above)
Comment (selfissued) [reply draft-jmiller-json-web-proof.md:121]: I disagree with the "multiple serializations to deserialize to the same buffer" comment.  There is a single base64url encoding of any string of bytes.  Given there's a 1:1 correspondence, I don't understand this comment, @OR13 .

**Review by selfissued — APPROVED** [review]
I believe that this PR moves us forward, and so should be merged.  That said, I'm fine with another commit being added to address some of the comments made in the PR beforehand.

File: draft-jmiller-json-web-proof.md (hunk 13 above)
Comment (OR13) [reply draft-jmiller-json-web-proof.md:121]: https://github.com/decentralized-identity/sidetree/blob/3184bfabd4b01e22c9d56fb16feadb5b303da9c4/lib/core/versions/1.0/Multihash.ts#L208

File: draft-jmiller-json-web-proof.md (hunk 12 above)
Comment (OR13) [reply draft-jmiller-json-web-proof.md:117]: is there a link to JSDT?

File: draft-jmiller-json-web-proof.md (hunk 6 above)
Comment (OR13) [reply draft-jmiller-json-web-proof.md:84]: I am in favor of:

```
type PayloadOptions = Buffer;
type Payload = base64url (Buffer). base64url (PayloadOptions).
type Payloads = Array<Payload>;
```

implementers shouldn't have to care about the payload content, the alg registry should tell you how to care about options or payloads.

File: draft-jmiller-json-web-proof.md (hunk 1 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:40]: Regardless of the cryptographic primitive used by the _issuer_, they are still performing the act of signing, the issuer is adding authentication (that they signed it) and integrity.  The holder upon receiving a _signed_ JWP then validates the signature to make sure it is authentic and has integrity.

When interacting with a verifier, the holder derives a proof of authenticity and integrity of the header and disclosed payloads.  The cryptographic primitive is predominately ZKPs, but the act is still proving since the holder isn't the original signer.

That's the way I've grown to think about it across all of the papers and code I've read in this space.  I think "proof generation" is more appropriate, as it is always an act of creation of a new output that is used to prove. Derivation could have other interpretations/meanings or create the wrong expectations.

File: draft-jmiller-json-web-proof.md (hunk 4 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:75]: I mentioned in another comment my thought process behind sign/sign_verify and proove/proof_verify language, happy to pick that up in this comment thread or we can create an issue to continue also.

Good point about the "only once" language, fixing that now :+1:

File: draft-jmiller-json-web-proof.md (hunk 6 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:85]: This is an unfortunate complexity I think we'll have to work through and deal with in some form.

The JWP _algorithm interface_ simply *must* support rich types.  JWP itself as a container format absolutely treats each payload as an array of bytes, but during the act of signing, proof generation, and proof validation the algorithm has to have the ability to accept rich types.

These rich interfaces are language/library level and not part of the protocol, but they are important to support required capabilities such as blinded values (algo has to be given an elliptic curve point), predicate proofs (algo requires a number, set member value, etc).

There's no reason to make these encodings be part of the protocol, they really are an implementation language level detail and algorithm specific.  There's not even an interop issue that I can tell since the algorithms that support these have very specific ways they use them via a hash2curve or other such cryptographic transformation.

I definitely need help explaining this and adding better text around it, so I welcome the discussions :)

File: draft-jmiller-json-web-proof.md (hunk 7 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:91]: I believe we'll be forming a sibling JWP spec that adds semantics around specific claims like JWT. Presentation exchange is also a good place for these types of rules.

I don't think these payload-specific requirements need to be handled at the core JWP layer other than integrity protecting and disclosing the fixed header.

File: draft-jmiller-json-web-proof.md (hunk 8 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:102]: Excellent point @OR13, this definitely needs some language around that, with probably more detail in a security considerations.

The protected header SHOULD be minimal and static across all usages of a given key, and static as in simply re-using the digested version.  It's only real purpose is to contain signals of how to process the JWP, key id, etc.

File: draft-jmiller-json-web-proof.md (hunk 9 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:108]: Agree that the payload contents should be addressed in a companion spec, and I also agree this wording is definitely incomplete and needs more work.

File: draft-jmiller-json-web-proof.md (hunk 10 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:110]: I'm going to close this sub-thread since it's the same discussion on the payload encodings above.

File: draft-jmiller-json-web-proof.md (hunk 12 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:120]: It is possible when the header digest is static and applies to all JWPs an issuer generates.

File: draft-jmiller-json-web-proof.md (hunk 13 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:124]: Given that we plan on defining a COSE Proof analog to JWP, my preference is to keep JWP as close to JOSE conventions as possible for easier adoption in those existing environments.

COSE Proofs can serve as the basis for much more efficient representations that can be serialized/encoded independently by a constrained transport as needed.

File: draft-jmiller-json-web-proof.md (hunk 14 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md]: Corrected.

File: draft-jmiller-json-web-proof.md (hunk 1 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:40]: I created issue #13 to continue this thread.

File: draft-jmiller-json-web-proof.md (hunk 6 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:85]: Let's continue this in issue #8

File: draft-jmiller-json-web-proof.md (hunk 9 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md]: Related to issue #8, I also made some minor improvements to this text to be more clear.

**Comment by quartzjer** [comment]
@OR13 I think I've either made some improvements or filed/linked issues for all the feedback and points you raised, if I missed any I'm happy to keep refining too.

File: draft-jmiller-json-web-proof.md (hunk 11 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md]: Pushed some improvements here that hopefully clarify also.

**Review by OR13 — APPROVED** [review]

**Comment by quartzjer** [comment]
Thanks all for the feedback!
